target_chains/starknet/contracts/src/pyth.cairo

```diff
@@ -1,29 +1,384 @@
+        self.owner.write(wormhole_address);
+        self.wormhole_address.write(wormhole_address);
+        write_data_sources(ref self, data_sources);
+    }
+
+    fn write_data_sources(ref self: ContractState, data_sources: Array<DataSource>) {
+        let num_old = self.num_data_sources.read();
+        let mut i = 0;
+        while i < num_old {
+            let old_source = self.data_sources.read(i);
+            self.is_valid_data_source.write(old_source, false);
+            self.data_sources.write(i, Default::default());
+            i += 1;
+        };
+
+        self.num_data_sources.write(data_sources.len());
+        i = 0;
+        while i < data_sources.len() {
+            let source = data_sources.at(i);
+            self.is_valid_data_source.write(*source, true);
+            self.data_sources.write(i, *source);
+            i += 1;
+        };
+    }
+
+    #[derive(Drop)]
+    struct PriceFeedMessage {
+        price_id: u256,
+        price: i64,
+        conf: u64,
+        expo: i32,
+        publish_time: u64,
+        prev_publish_time: u64,
+        ema_price: i64,
+        ema_conf: u64,
     }
 
     #[abi(embed_v0)]
-    impl HelloImpl of super::IPyth<ContractState> {
-        fn get_name(self: @ContractState) -> felt252 {
-            self.name.read() + 1
+    impl PythImpl of super::IPyth<ContractState> {
+        fn latest_price_info(self: @ContractState, price_id: u256) -> PriceInfo {
+            self.latest_price_info.read(price_id)
         }
 
-        fn set_name(ref self: ContractState, name: felt252) {
-            self.name.write(name - 1);
+        fn set_data_sources(
+            ref self: ContractState, sources: Array<DataSource>
+        ) -> Result<(), SetDataSourcesError> {
+            if self.owner.read() != get_caller_address() {
+                return Result::Err(SetDataSourcesError::AccessDenied);
+            }
+            write_data_sources(ref self, sources);
+            Result::Ok(())
+        }
+
+        fn update_price_feeds(
+            ref self: ContractState, data: ByteArray
+        ) -> Result<(), UpdatePriceFeedsError> {
+            let mut reader = ReaderImpl::new(data);
+            let x = reader.read_u32().map_err()?;
+            if x != ACCUMULATOR_MAGIC {
+                return Result::Err(UpdatePriceFeedsError::InvalidUpdateData);
+            }
+            if reader.read_u8().map_err()? != MAJOR_VERSION {
+                return Result::Err(UpdatePriceFeedsError::InvalidUpdateData);
+            }
+            if reader.read_u8().map_err()? < MINIMUM_ALLOWED_MINOR_VERSION {
+                return Result::Err(UpdatePriceFeedsError::InvalidUpdateData);
+            }
+
+            let trailing_header_size = reader.read_u8().map_err()?;
+            reader.skip(trailing_header_size).map_err()?;
+
+            let update_type: Option<UpdateType> = reader.read_u8().map_err()?.try_into();
+            match update_type {
+                Option::Some(_) => {},
```

Comment:
I think matching specifically here would be less error prone, match on `Option::Some(WormholeMerkle)`, and else on `_`

Reply:
Added an explicit match here and in similar places. Unfortunately, nested patterns like `Option::Some(WormholeMerkle)` are not supported so it's just a nested match.